### ISSUE TITLE
[Core] Ray appears to be leaking IDLE workers, even after their jobs have finished

### ISSUE THREAD

**Issue by jfaust** [ray-project/ray#44438]
### What happened + What you expected to happen

We have a pipeline that is running multiple jobs in parallel through a cluster with a single node. We noticed memory usage increasing over time, and I'm in the process of investigating.

What I've found is that Ray seems to be leaking IDLE worker nodes, and each node is using between 300 and 500MB of RAM. When we first discovered the issue, there were ~39 IDLE workers taking up about 20GB of memory.  Eventually processes start getting OOM killed.

Running locally on my machine with some test data, after about an hour I see ~25 IDLE workers that do not go away after their jobs are finished. Here's a screenshot from the dashboard (this is just a subset of the IDLE workers):

<img width="944" alt="Screenshot 2024-04-02 162332" src="https://github.com/ray-project/ray/assets/1388377/235237c4-d25e-424f-b1aa-30384dd03239">

Here I've run all the commands I think might be useful:
```bash
$ ray status      
======== Autoscaler status: 2024-04-02 15:59:02.835618 ========
Node status
---------------------------------------------------------------
Active:
 1 node_5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1
Pending:
 (no pending nodes)
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------
Usage:
 0.0/16.0 CPU
 0.0/1.0 GPU
 0.0/8589934592.0 gpu_memory
 0B/25.64GiB memory
 0B/12.82GiB object_store_memory

Demands:
 (no resource demands)


$ ray list workers -f IS_ALIVE=True

======== List: 2024-04-02 15:48:07.858153 ========
Stats:
------------------------------
Total: 34

Table:
------------------------------
    WORKER_ID                                                 IS_ALIVE    WORKER_TYPE     EXIT_TYPE    NODE_ID                                                   IP                PID
 0  02000000ffffffffffffffffffffffffffffffffffffffffffffffff  True        DRIVER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   8346
 1  041a44975eb4f56538ef4dd056ea1471e7170ba2a0dac31300b961fd  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  27997
 2  274237149f07287c2bdd92052ed7dadff2d8629966550456b3f9e2c6  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  11793
 3  274d5105c80fbd0a15220863a046df5a8374d06ddfdd12bf5dd7994a  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  17692
 4  284a3e4753b59555e6cc906923c829a6057b092711ba30f5bd5048a4  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  14007
 5  28ca2f81698a2df62db7a3e48256b5869e914ed0aa7728f401f4c203  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  11562
 6  29e79468f76305d2a3b2c78e9311c39ac395f9f16f68b7a800eb0f0f  True        SPILL_WORKER                 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10071
 7  2ae42f8227c2a29087078d7913fbdf52e0fc343fb86c5697f327a9c8  True        RESTORE_WORKER               5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10667
 8  327b1f6167215440cb1c166fb791967ac5dcd41505ba891e76587a74  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   3593
 9  38b22712d5f19791864cb5c67ac64ff0c7b6558735d6db0de056379e  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  19921
10  38b8d7d7c1e8ddd4fe8c663cf07ce96eb4c6acaac564aca3b0f64567  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   7371
11  420a1e0b81715e14a19ca6e02331b7cd49071a53d42d3cb6a4a81dc5  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   7942
12  44d279e56befd61e8eb603d51904af318572a983951691875d991ad5  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  14949
13  45cab92cbbe76c753ab9417dcb882ce832be42c8ad085998e398afe0  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  26411
14  4ca6b29344562a111f66af3fdbb8235723435aa2149ca8d0981de7fa  True        SPILL_WORKER                 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10026
15  64e286c84ddce4625ff3de60763ce5a9fadbcc939d40d32b82732c12  True        RESTORE_WORKER               5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10592
16  6948d734a6a47eca93f2a3161ef8466f1938d343d3820d7dca424711  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  14985
17  6b1206337bc4df09a14094e291fb4bdacbe47f9d96cada4d98af12a6  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   6947
18  7a4d146e2758148282b745bd9978796f5620cab08cb549d170258ccb  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  23179
19  8099b08523fe63f0cc48421dd159521dcf34dec0cdb2aa6b5a0ccff3  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   6338
20  93a2894e5b4ec9bb265a95b5667e34dfce0ba4d61924a2755afdc0f0  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  30510
21  a662749919643a6c76cb215b07f6d3fd51653b46330e79c37f58cd82  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  11586
22  a7f4a5a9ba2e20aca04e9023bce662e3f0fcc4ed6399a188b1f1e2c2  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  32720
23  af163d2c9bd29d11771b1ff865bd1b1a13b213e6edcce99cdc35b884  True        SPILL_WORKER                 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10072
24  bb5a9a9888acc1b75bb0db2cdde105d0278289f3eefee24262adfeab  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  24514
25  c07027cf2528ed6ba48f88d44205659ce584a48701ef5ab3c02ce341  True        RESTORE_WORKER               5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10737
26  c2af2f3ca26328105aec2f8a6b779baca0837edcb58f10f93fb85a6d  True        SPILL_WORKER                 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10058
27  ccd9b51746dc4bb70149e735a63f94e9e30f1937d06498663b8a0d8f  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   2603
28  ce5120360e797c7c0c81221ad7152a5a76fb509e48099c3f0caf8cd1  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  30906
29  d9c0cd7aed477127472ec438be3ec0bd5fcff353d23d3ef9e8c350cb  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   6748
30  dec9de7b2966a24fc8d6004c28737f7cb28a42ec723feea2d79d6d79  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   4612
31  df91767d52952c27c9684f23271e6b7ac1b277d7d3dbacf4e476a144  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  32352
32  e57a121a1f721689a9649c1ec27d5e934132fd71d0f0234f73fef4e7  True        RESTORE_WORKER               5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127  10666
33  ea66291d1bc00f980a4f2e3c08e2fc7626eaa8bab91648f4e52b8a3b  True        WORKER                       5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1  172.23.208.127   9437

$ ray memory       
======== Object references status: 2024-04-02 16:00:19.819896 ========
Grouping by node address...        Sorting by object size...        Display allentries per group...


To record callsite information for each ObjectRef created, set env variable RAY_record_ref_creation_sites=1

--- Aggregate object store stats across all nodes ---
Plasma memory usage 0 MiB, 0 objects, 0.0% full, 0.0% needed
Spilled 471006 MiB, 694 objects, avg write throughput 1137 MiB/s
Restored 121952 MiB, 249 objects, avg read throughput 69 MiB/s
Objects consumed by Ray tasks: 2043111 MiB.

$ ray list objects
No resource in the cluster

$ ray list jobs -f 'STATUS!=SUCCEEDED'
No resource in the cluster

$ ray list tasks -f 'STATE!=FINISHED'
No resource in the cluster
```

So, as far as Ray is concerned, as far as I can tell:
1. There are no jobs running
2. There is nothing in memory, and no objects in the object store
3. Yet these workers stay resident

I'm trying to create a minimal repro, but have so far been unable to. I'm hoping maybe someone has seen something similar and can point me in the right direction (this seems similar to #31967 or #25699, except there does not appear to be anything pinned in memory preventing those workers from being killed).

Looking at the worker logs, it has been told to exit but never does (this may also be an Actor worker? I could be misreading this):
```
143[2024-04-02 15:00:09,821 I 2603 2723] actor_manager.cc:214: received notification on actor, state: ALIVE, actor_id: 227fb55653696d7910c8326d35000000, ip address: 172.23.208.127, port: 10655, worker_id: b9ecb7db64de0b332162fc7c18b999cc7d72e8b510f0023e61d931f8, raylet_id: 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1, num_restarts: 0, death context type=CONTEXT_NOT_SET
144[2024-04-02 15:00:09,821 I 2603 2723] direct_actor_task_submitter.cc:237: Connecting to actor 227fb55653696d7910c8326d35000000 at worker b9ecb7db64de0b332162fc7c18b999cc7d72e8b510f0023e61d931f8
145[2024-04-02 15:00:10,273 I 2603 2723] actor_manager.cc:214: received notification on actor, state: DEAD, actor_id: 227fb55653696d7910c8326d35000000, ip address: 172.23.208.127, port: 10655, worker_id: b9ecb7db64de0b332162fc7c18b999cc7d72e8b510f0023e61d931f8, raylet_id: 5057ba6e230e96eb323ee6656286351abb9e97a555f36dc0afc3e4d1, num_restarts: 0, death context type=ActorDiedErrorContext
146[2024-04-02 15:00:10,274 I 2603 2723] direct_actor_task_submitter.cc:287: Failing pending tasks for actor 227fb55653696d7910c8326d35000000 because the actor is already dead.
147[2024-04-02 15:00:10,408 I 2603 2723] core_worker.cc:4197: Force exiting worker that owns object. This may cause other workers that depends on the object to lose it. Own objects: 1 # Pins in flight: 0
148[2024-04-02 15:00:10,409 I 2603 2723] core_worker.cc:835: Exit signal received, this process will exit after all outstanding tasks have finished, exit_type=INTENDED_SYSTEM_EXIT, detail=Worker exits because it was idle (it doesn't have objects it owns while no task or actor has been scheduled) for a long time.
149[2024-04-02 15:00:10,409 W 2603 2723] task_manager.cc:387: This worker is still managing 1 in flight tasks, waiting for them to finish before shutting down.
```
The logs then just continue with their periodic dumping of stats, and the worker never exits.
Every one of the leaked workers has a log similar to this one.

Any thoughts on what would be preventing these workers from exiting?

### Versions / Dependencies

* Tried Ray 2.9.2, Ray 2.10.0 
* Python 3.8
* Ubuntu 20.04

### Reproduction script

I've been unfortunately unable to create a minimal repro so far. I will update here if I manage to.

### Issue Severity

High: It blocks me from completing my task.

**Comment by jfaust** [comment]
Even with the Actor referenced in the above log completely removed, I still get:
```
127[2024-04-02 17:41:53,187 I 3724 3858] core_worker.cc:4197: Force exiting worker that owns object. This may cause other workers that depends on the object to lose it. Own objects: 1 # Pins in flight: 0
128[2024-04-02 17:41:53,187 I 3724 3858] core_worker.cc:835: Exit signal received, this process will exit after all outstanding tasks have finished, exit_type=INTENDED_SYSTEM_EXIT, detail=Worker exits because it was idle (it doesn't have objects it owns while no task or actor has been scheduled) for a long time.
129[2024-04-02 17:41:53,187 W 3724 3724] reference_count.cc:54: This worker is still managing 1 objects, waiting for them to go out of scope before shutting down.
```

And the worker never exits, so the Actor bit may be a red herring.

**Comment by jfaust** [comment]
A little more information, from looking at different logs. It appears that the worker that never exits is for the very top level task in our pipeline - it's a task that just submits a bunch of other tasks to run and then waits for results.

**Comment by jfaust** [comment]
After compiling Ray from source and enabling debug logging in the core, I found a repro (added to the initial comment).

**Comment by jfaust** [comment]
Looks like maybe it doesn't need to be returned to the Driver. This seems to reproduce as well:
```python
import ray
import numpy as np

ray.init(namespace="test")


@ray.remote(name="actor", get_if_exists=True, lifetime="detached")
class Actor:

    def do_work(self):
        arr = np.zeros((10000, 1000))
        np.multiply(arr, arr)


@ray.remote
def _create_actor():
    return Actor.remote()


@ray.remote
def _toplevel():
    actor = ray.get(_create_actor.remote())
    tasks = [actor.do_work.remote() for _ in range(0, 100)]
    ray.get(tasks)


ray.get(_toplevel.remote())
```

**Comment by bug-catcher** [comment]
@jfaust what is the workaround? this is causing a lot of issues on our cluster

**Comment by jfaust** [comment]
> @jfaust what is the workaround? this is causing a lot of issues on our cluster

@bug-catcher in my case it was easy to create the Actor and pass it places and never return it. Not a workaround in many cases but worked for us.

**Comment by vipese-idoven** [comment]
Are there any updates on this?

I'm running into a similar issue here. In our case, the detached Ray Actor is used to pre-process a Ray Dataset. If the job fails, stops, or succeeds, several IDLE and IDLE_Spill workers remain allocating the majority of the memory.

**Comment by jjyao** [comment]
> I'm running into a similar issue here. In our case, the detached Ray Actor is used to pre-process a Ray Dataset. If the job fails, stops, or succeeds, several IDLE and IDLE_Spill workers remain allocating the majority of the memory.

@vipese-idoven, if you are using detached actor, then after the Job finishes, detached actor will still be running and all the workers created by it will not be force killed.

**Comment by vipese-idoven** [comment]
> > I'm running into a similar issue here. In our case, the detached Ray Actor is used to pre-process a Ray Dataset. If the job fails, stops, or succeeds, several IDLE and IDLE_Spill workers remain allocating the majority of the memory.
> 
> @vipese-idoven, if you are using detached actor, then after the Job finishes, detached actor will still be running and all the workers created by it will not be force killed.

Thanks for the quick response! Just to confirm: when a detached Actor triggers the creation of workers like ray::IDLE and ray::IDLE_spill workers, these will not be force kill. From what I've tried, killing the detached Actor does not kill them either. Is there a way to programmatically kill those workers that are apparently unused by other Tasks/Actors?
